### PR TITLE
feat: add tools whois

### PR DESCRIPTION
**Pull request by Rigonkmalk**

### CODE REVIEW

File: app/models/whois.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class Whois < ApplicationRecord
+  validates :domain, presence: true
```
Comment (sebastiendeu) [inline app/models/whois.rb:4]: Why you have created a model ?

File: app/views/utils/whois.html.haml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      - if @whois.property_any_supported? :nameservers
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Nameserver')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }
+            - @whois.nameservers.each do |nameserver|
+              %dl
+              = nameserver.name
+
+
+      - if @whois.property_any_supported? :nameservers
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:20]: You can use last condition to reduce code verbosity

File: app/views/utils/whois.html.haml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('IPv6')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.nameservers.first.ipv6
+
+      %div{ class: 'bg-white dark:bg-gray-600 my-8 shadow overflow-hidden sm:rounded-lg' }
+        %div{ class: 'px-4 py-5 sm:px-6' }
+          %h3{ class: 'text-lg leading-6 font-medium text-gray-900 dark:text-gray-200' }= _('Contacts')
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:34]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('ID')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.id
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:39]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Type')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.type
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:44]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Name')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.name
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:49]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Organization')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.organization
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:54]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Address')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.address
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:59]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('City')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.city
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:64]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Zip')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.zip
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:69]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('State')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.state
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:74]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Country')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.country
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:79]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Country Code')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.country_code
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:84]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Phone')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.phone
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:89]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Fax')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.fax
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:94]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Email')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.email
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:99]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Phone')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.fax
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:104]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('URL')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.url
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:109]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Created on')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.contacts.first.created_on
+
+      - if @whois
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:114]: You have already condition on @whois

File: app/views/utils/whois.html.haml [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+              %dl
+              = nameserver.name
+
+
+      - if @whois.property_any_supported? :nameservers
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('IPv4')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.nameservers.first.ipv4
+
+      - if @whois.property_any_supported? :nameservers
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:25]: You can use last condition to reduce code verbosity

File: app/views/utils/whois.html.haml [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+      %div{ class: 'bg-white dark:bg-gray-600 my-8 shadow overflow-hidden sm:rounded-lg' }
+        %div{ class: 'px-4 py-5 sm:px-6' }
+          %h3{ class: 'text-lg leading-6 font-medium text-gray-900 dark:text-gray-200' }= _('Registrar')
+
+      - if @whois.property_any_supported? :registrar
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('ID')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.registrar.id
+
+      - if @whois.property_any_supported? :registrar
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:128]: You can use last condition to reduce code verbosity

File: app/views/utils/whois.html.haml [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-gray-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Name')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.registrar.name
+
+      - if @whois.property_any_supported? :registrar
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:133]: You can use last condition to reduce code verbosity

File: app/views/utils/whois.html.haml [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+        %div{ class: 'bg-white-50 dark:bg-gray-500 px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6' }
+          %dt{ class: 'text-sm font-medium text-gray-500 dark:text-gray-300' }= _('Organization')
+          %dd{ class: 'mt-1 text-sm text-gray-900 dark:text-gray-100 sm:mt-0 sm:col-span-2' }= @whois.registrar.organization
+
+      - if @whois.property_any_supported? :registrar
```
Comment (sebastiendeu) [inline app/views/utils/whois.html.haml:138]: You can use last condition to reduce code verbosity

**Review by sebastiendeu — CHANGES_REQUESTED** [review]

**Review by vfricou — APPROVED** [review]

**Review by sebastiendeu — APPROVED** [review]

**Pull request merged**